pipeline_tools/create_envelope.py

```diff
@@ -1,277 +1,361 @@
 
     Returns:
-        uuid (str): A string representing the input bundle uuid
+        output_refs (List[dict]): A list of metadata dicts describing the output files produced by the analysis_process.
     """
-    bundle = analysis_json['input_bundles'][0]
-    uuid = bundle
-    print('Input bundle uuid {0}'.format(uuid))
-    return uuid
+    outputs = analysis_process['outputs']
+    output_refs = [
+        {
+            'fileName': out['file_core']['file_name'],
+            'content' : {
+                'describedBy': '{0}/type/file/{1}/analysis_file'.format(raw_schema_url, analysis_file_version),
+                'schema_type': 'file',
+                'file_core'  : {
+                    'file_name'  : out['file_core']['file_name'],
+                    'file_format': out['file_core']['file_format']
+                }
+            }
+        } for out in outputs
+    ]
+    return output_refs
 
 
-def get_output_files(analysis_json, schema_url, analysis_file_version):
-    """Get the metadata describing the outputs of the analysis
+def add_file_reference(file_ref, file_refs_url, auth_headers, http_requests):
+    """Add a file reference to the analysis metadata in a submission envelope.
 
     Args:
-        analysis_json (dict): metadata describing the analysis
-        schema_url (str): URL for retrieving HCA metadata schemas
-        analysis_file_version (str): the schema version that file references will conform to
+        file_ref (dict): HCA metadata stub about the file.
+        file_refs_url (str): The url for adding file references
+        auth_headers (dict): Dict representing headers to use for auth.
+        http_requests (http_requests.HttpRequests): The HttpRequests object to use for talking to Ingest.
 
     Returns:
-        output_refs (dict): A dict of metadata describing the output files produced by the analysis
+        dict: Dict representing the JSON response to the request that adding the file reference.
+
+    Raises:
+        requests.HTTPError: For 4xx errors or 5xx errors beyond timeout.
     """
-    outputs = analysis_json['outputs']
-    output_refs = []
-
-    for out in outputs:
-        output_ref = {}
-        file_name = out['file_core']['file_name']
-        output_ref['fileName'] = file_name
-        output_ref['content'] = {
-            'describedBy': '{}/type/file/{}/analysis_file'.format(schema_url, analysis_file_version),
-            'schema_type': 'file',
-            'file_core': {
-                'file_name': file_name,
-                'file_format': out['file_core']['file_format']
-            }
-        }
-        output_refs.append(output_ref)
-    return output_refs
+    print('Adding file: {0} to the file reference.'.format(file_ref['fileName']))
+    response = http_requests.put(file_refs_url, headers=auth_headers, json=file_ref)
+    return response.json()
+
+
+def build_envelope(submit_url, analysis_protocol_path, analysis_process_path, raw_schema_url,
```

Comment:
I find it easier to read when the top-level function (`build_envelope` in this case), is at the top of the file, so that the first thing I read is more of an overview, and then I can scroll down to see the details / helper functions.

Reply:
Yes, that totally makes sense, I wrote this function at last during development, will push it to the top!